src/client.rs

```diff
@@ -1,107 +1,132 @@
 use base64::{engine::general_purpose, Engine as _};
+use bytes::Bytes;
+use etherparse::Ipv4Header;
+use futures::stream::SplitSink;
 use futures::stream::StreamExt;
-use rand::distributions::Standard;
+use futures::SinkExt;
+use ipnetwork::Ipv4Network;
+use rand::distributions::Alphanumeric;
 use rand::{thread_rng, Rng};
-use std::io::Read;
+use std::io::{Cursor, Write};
+use std::net::Ipv4Addr;
 use std::sync::Arc;
 use tokio::{sync::mpsc, task};
+use tokio_util::codec::Framed;
+use tun::AsyncDevice;
+use tun::TunPacketCodec;
 use tun::{IntoAddress as _, TunPacket};
 
 use crate::config;
 use crate::remote;
 
+type TunSink = SplitSink<Framed<AsyncDevice, TunPacketCodec>, TunPacket>;
+
 pub struct Client {
-    tunnels: Vec<Arc<Tunnel>>,
-    receiver: mpsc::Receiver<TunPacket>,
-    remote: Arc<remote::Remote>,
+    tunnels: Arc<Vec<Tunnel>>,
+    remote_receiver: mpsc::Receiver<(String, Bytes)>,
+    local_addr: Ipv4Addr,
+    sink: TunSink,
 }
 
-#[derive(Clone)]
 struct Tunnel {
     id: Vec<u8>,
     topic: String,
+    topic_base: String,
     bind_addr: std::net::Ipv4Addr,
 }
 
 impl Client {
     pub async fn new(config: &config::Config) -> Self {
+        log::info!(
+            "Creating tun at {:?} network {:?}",
+            config.driver.local_addr,
+            config.driver.tun.netmask
+        );
+        let mut tun_config = tun::Configuration::default();
+        tun_config.address(config.driver.local_addr);
+        tun_config.destination(config.driver.local_addr);
+        tun_config.netmask(config.driver.tun.netmask);
```

Comment:
I think we might need to rethink the config structure...

Server-side can derrive everything from a single subnet in cidr notation.

Client-side can also have multiple tunnels that must be in the same subnet but I think each tunnel's IP should be configurable eg:
```
[
    { topic: "", addr: "" },
    ...
]
```
Maybe the client-side `addr` can also be in cidr notation and we verify that all of them belongs to the same subnet + 1st IP is reserved for yourself.